discretize/TensorMesh.py

```diff
@@ -18,25 +19,58 @@ class BaseTensorMesh(BaseMesh):
 
     _unitDimensions = [1, 1, 1]
 
-    def __init__(self, h_in, x0_in=None):
-        assert type(h_in) in [list, tuple], 'h_in must be a list'
-        assert len(h_in) in [1, 2, 3], 'h_in must be of dimension 1, 2, or 3'
-        h = list(range(len(h_in)))
-        for i, h_i in enumerate(h_in):
-            if utils.isScalar(h_i) and type(h_i) is not np.ndarray:
-                # This gives you something over the unit cube.
-                h_i = self._unitDimensions[i] * np.ones(int(h_i))/int(h_i)
-            elif type(h_i) is list:
-                h_i = utils.meshTensor(h_i)
-            assert isinstance(h_i, np.ndarray), (
-                "h[{0:d}] is not a numpy array.".format(i)
+    # properties
+    h = properties.List(
+        "h is a list containing the cell widths of the tensor mesh in each "
+        "dimension.",
+        properties.Array(
+            "widths of the tensor mesh in a single dimension",
+            dtype=float,
+            shape=("*",)
+        ),
+        max_length=3
+    )
+
+    def __init__(self, h=None, x0=None, **kwargs):
+
+        h_in = h
+        x0_in = x0
+
+        # Cell widths
+        if h_in is not None:
+            # Sanity Checks
+            assert 'h' not in kwargs.keys(), (
+                'Only 1 descriptor of cell sizes can be used to instantiate '
+                'the class'
             )
-            assert len(h_i.shape) == 1, (
-                "h[{0:d}] must be a 1D numpy array.".format(i)
+            assert type(h_in) in [list, tuple], 'h_in must be a list'
+            assert len(h_in) in [1, 2, 3], (
+                'h_in must be of dimension 1, 2, or 3'
             )
-            h[i] = h_i[:]  # make a copy.
 
+            # build h
+            h = list(range(len(h_in)))
+            for i, h_i in enumerate(h_in):
+                if utils.isScalar(h_i) and type(h_i) is not np.ndarray:
+                    # This gives you something over the unit cube.
+                    h_i = self._unitDimensions[i] * np.ones(int(h_i))/int(h_i)
+                elif type(h_i) is list:
+                    h_i = utils.meshTensor(h_i)
+                assert isinstance(h_i, np.ndarray), (
+                    "h[{0:d}] is not a numpy array.".format(i)
+                )
+                assert len(h_i.shape) == 1, (
+                    "h[{0:d}] must be a 1D numpy array.".format(i)
+                )
+                h[i] = h_i[:]  # make a copy.
+
+        else:
+            assert 'h' in kwargs.keys(), 'cell widths must be provided'
+            h = kwargs.pop('h')
```

Comment:
This is a little funny - on first impulse, my question was "Why does `h` get special treatment as an arg but not as a kwarg?" Looking into it further, though, as long as h is provided as an arg or a kwarg, it will set `h_in`. So really, this line will never be reached; if the `else` clause is reached, the `assert` will fail (I think at least).

Silly example:
![image](https://user-images.githubusercontent.com/9453731/30046713-0532993a-91ca-11e7-9bb9-2fb7254c1974.png)

Reply:
It is a bit odd, and maybe you have a suggestion for me on how this might be made more transparent?

The reason for this is to support two uses: 
1. creating the mesh with `h, x0` as args (or `h_in`). This is currently how most people work with the API
2. the reason for having `h` also allowed as a kwarg is so that I can create a re-create a mesh from a serialized mesh

![image](https://user-images.githubusercontent.com/6361812/30069521-cad0351c-9215-11e7-932d-28a1c4ab5901.png)

where `load_mesh` deserializes the mesh (https://github.com/simpeg/discretize/blob/ref/properties/discretize/utils/meshutils.py#L17)

```python
def load_mesh(filename):
    """
    Open a json file and load the mesh into the target class
    As long as there are no namespace conflicts, the target __class__
    will be stored on the properties.HasProperties registry and may be
    fetched from there.
    :param str filename: name of file to read in
    """
    with open(filename, 'r') as outfile:
        jsondict = json.load(outfile)
        data = properties.HasProperties.deserialize(jsondict, trusted=True)
    return data
``` 

and for this I am under the impression that all properties in the serialized json needed to be able to be read in as kwargs.

Reply:
I see the need for h as both arg and kwarg - I think that is still covered.

This exposes a bit of a subtlety with how Python treats args/kwargs. If a named arg variable shows up anywhere in the kwargs (and is not provided in the args) it is assigned to the associated arg (and that variable does not show up in `kwargs` internal to the function).

If you play around with this:

```python
class Blah(object):
    
    def __init__(self, h=None, x0=None, **kwargs):
        print(h)
        print(x0)
        print(kwargs)
```

You will find for all of these, `kwargs = {'something_else': 2}`:

```python
Blah(0, 1, something_else=2)
Blah(h=0, x0=1, something_else=2)
Blah(something_else=2, x0=1, h=0)
```
This errors (h assigned twice):
```python
Blah(0, x0=1, something_else=2, h=3)
```

So bottom line, even if it is coming in as a deserialized dict, the argument will still be moved to the right place. Does that make sense...?